compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp

```diff
@@ -83,13 +86,81 @@ inferVectorSizesFromIR(linalg::LinalgOp linalgOp) {
   return vectorSizes;
 }
 
-// Return the vector sizes from the local lowering config or try to infer them
+static std::optional<SmallVector<int64_t>>
+inferVectorSizesFromIR(tensor::PackOp op) {
+  LLVM_DEBUG(VEC_DBGS() << "Inferring vector sizes for:\n" << op << "\n");
+
+  if (llvm::any_of(op.getInnerTiles(), [](OpFoldResult v) {
+        return !getConstantIntValue(v).has_value();
+      })) {
+    LLVM_DEBUG(VEC_DBGS() << "skip, because inner_tiles are not all constant");
+    return std::nullopt;
+  }
+
+  SmallVector<int64_t> vectorSizes;
+  auto inferred = inferVectorSizesFromIR(op.getSource());
+  if (!inferred) {
+    return std::nullopt;
+  }
+  vectorSizes = inferred.value();
+
+  for (auto [dimPos, tileSize] :
+       llvm::zip_equal(op.getInnerDimsPos(), op.getStaticInnerTiles())) {
+    if (vectorSizes[dimPos] % tileSize != 0) {
+      return std::nullopt;
+    }
+    vectorSizes[dimPos] /= tileSize;
+  }
+  auto outerDimsPerm = op.getOuterDimsPerm();
+  if (!outerDimsPerm.empty()) {
+    applyPermutationToVector(vectorSizes, outerDimsPerm);
+  }
+
+  LLVM_DEBUG({
+    VEC_DBGS() << "After adjustment with inner tiles and "
+                  "outer_dims_perm:\n";
+    for (auto [idx, val] : llvm::enumerate(vectorSizes)) {
+      llvm::dbgs() << "Dim #" << idx << ": " << val << "\n";
+    }
+  });
+
+  return vectorSizes;
+}
+
+static std::optional<SmallVector<int64_t>> inferVectorSizesFromIR(Value val) {
+  std::optional<SmallVector<int64_t>> result;
+  TypeSwitch<Operation *, void>(val.getDefiningOp())
+      .Case<linalg::LinalgOp, tensor::PackOp>(
+          [&](auto op) { result = inferVectorSizesFromIR(op); })
+      .Case<tensor::ExtractSliceOp>([&](tensor::ExtractSliceOp op) {
```

Comment:
Since `inferVectorSizesFromIR` takes `Operation *` should we just make the default case do `result = inferVectorSizesFromIR(op);` and remove the case for `linalg::LinalgOp, tensor::PackOp`?

This would be avoiding inferring vector sizes from a `tensor::PadOp`. Is there a reason to avoid that case when inferring through value chains?

Reply:
There are missing functionality for pad op vectorization. Actually there are few vectorization patterns. Some of them live in IREE and some of them live in upstream. It is quite complicated even just with upstream patterns. They have different behaviors with maskings. I'd like to clean these up for a long time, but it is never on people's priorities because so far we don't fuse pad op by default. Unfortunately it is needed to maintain default and experimental paths...